src/device/socket/vsock.rs

```diff
@@ -101,24 +139,185 @@ impl<'a, H: Hal, T: Transport> VirtIOSocket<'a, H, T> {
             op: VirtioVsockOp::Request.into(),
             ..Default::default()
         };
-        self.tx
-            .add_notify_wait_pop(&[header.as_bytes(), &[]], &mut [], &mut self.transport)?;
+        self.send_packet_to_tx_queue(&header, &[])?;
+        let received_header = self.pop_header_only_packet_from_rx_queue()?;
+        match received_header.op()? {
+            VirtioVsockOp::Response => {
+                let dst = VsockAddr {
+                    cid: dst_cid,
+                    port: dst_port,
+                };
+                self.connection_info = Some(ConnectionInfo {
+                    dst,
+                    src_port,
+                    ..Default::default()
+                });
+                Ok(())
+            }
+            VirtioVsockOp::Rst => Err(SocketError::ConnectionFailed.into()),
+            VirtioVsockOp::Shutdown => Err(SocketError::PeerSocketShutdown.into()),
+            _ => Err(SocketError::InvalidOperation.into()),
+        }
+    }
+
+    /// Requests the credit and updates the credit in the current connection info.
+    pub fn request_credit(&mut self) -> Result {
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::CreditRequest.into(),
+            ..Default::default()
+        };
+        self.send_packet_to_tx_queue(&header, &[])?;
+        self.wait_one_in_rx_queue();
+        let received_header = self.pop_header_only_packet_from_rx_queue()?;
+        match received_header.op()? {
+            VirtioVsockOp::CreditUpdate => {
+                if let Some(connection_info) = self.connection_info.as_mut() {
+                    connection_info.buf_alloc = received_header.buf_alloc.into();
+                    connection_info.fwd_cnt = received_header.fwd_cnt.into();
+                }
+                debug!("Connection info {:?}", self.connection_info);
+                Ok(())
+            }
+            VirtioVsockOp::Rst => Err(SocketError::ConnectionFailed.into()),
+            _ => Err(SocketError::InvalidOperation.into()),
+        }
+    }
+
+    /// Sends the buffer to the destination.
+    pub fn send(&mut self, buffer: &[u8]) -> Result {
+        self.request_credit()?;
```

Comment:
Why do you need to call this every time before sending? I'm not super familiar with the details of vsock, but can't you check the free space you already know about first? And if there is no space available, I guess you shouldn't send the packet yet?